theano/gpuarray/opt.py

```diff
@@ -215,6 +256,248 @@ def apply(self, fgraph):
+                    node_created.setdefault(lopt, 0)
+
+        for node in topo:
+
+            if isinstance(node.op, HostFromGpu):
+                mapping[node.outputs[0]] = node.inputs[0]
+                continue
+
+            # Move only if any of the inputs are on the GPU.
+            move_to_GPU = False
+            from .type import GpuArrayVariable, GpuArraySharedVariable #when you uncomment
+            if any([isinstance(i, GpuArrayVariable) or
+                   isinstance(i, GpuArraySharedVariable)
+                   for i in [mapping[v] for v in node.inputs] +
+                   node.outputs]):
+
+                move_to_GPU = True
+
+            context_name = None
+            for i in [mapping[i] for i in node.inputs]:
+                if isinstance(i.type, GpuArrayType):
+                    context_name = i.type.context_name
+                    move_to_GPU = True
+                    break
+            if (not move_to_GPU and
+                    isinstance(node.op, (theano.tensor.Alloc,
+                                         theano.tensor.AllocEmpty))):
+                # If the Alloc[Empty] have a client that will be moved
+                # to the GPU, we should move the Alloc* on the GPU.
+
+                # We approximate this by supposing that if we have an
+                # optimization for one of the clients op, then we will
+                # move the client to the GPU.
+                for c, _ in node.outputs[0].clients:
+                    if (c != 'output' and
+                        (self.local_optimizers_map.get(c.op, []) +
+                         self.local_optimizers_map.get(type(c.op), []))):
+                        move_to_GPU = True
+            new_ops = None
+            outputs = []
+            # Apply the lifter
+            for lopt in (self.local_optimizers_map.get(node.op, []) +
+                         self.local_optimizers_map.get(type(node.op), []) +
+                         self.local_optimizers_all):
+                if move_to_GPU:
+                    t_opt = time.time()
+                    new_ops = lopt.transform(node.op, context_name,
+                                             [mapping[i] for i in node.inputs],
+                                             node.outputs)
+                    t_opt2 = time.time()
+                    time_opts[lopt] += t_opt2 - t_opt
+
+                    if new_ops:
+                        process_count[lopt] += 1
+                        break
+            if not new_ops:
+                newnode = node.clone_with_new_inputs([mapping.get(i)
+                                                      for i in node.inputs])
+                new_not_transferred.append(newnode)
+
+                outputs = newnode.outputs
+            elif isinstance(new_ops, (tuple, list)):
+                outputs = []
+                for o in new_ops:
+                    if o.owner and isinstance(o.owner.op, HostFromGpu):
```

Comment:
We removed it from the opt at few places. So yeah I'll do for the remaining as well